README.md
Android Sliding Up Panel
=========================


![SlidingUpPanelLayout](https://raw.github.com/josomers/SlidingUpPanelLayout/master/slidinguppanel.png)

Usage
-----------
To use the layout, simply include `com.slidinguppanel.SlidingUpPanelLayout` as the root element in your activity Layout. Make sure that it has two children. The first child is your main layout. The second child is your layout for the sliding up panel. Both children should have width and height set to `match_parent`. For more information, please refer to the sample code.

```xml
<com.slidinguppanel.SlidingUpPanelLayout
    android:id="@+id/sliding_layout"
    android:layout_width="match_parent"
    android:layout_height="match_parent" >

    <TextView
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:gravity="center"
        android:text="Main Content"
        android:textSize="16sp" />

    <TextView
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:gravity="center|top"
        android:text="The Awesome Sliding Up Panel"
        android:textSize="16sp" />
</com.slidinguppanel.SlidingUpPanelLayout>
```
For smooth interaction with the ActionBar, make sure that `windowActionBarOverlay` is set to `true` in your styles:
```xml
<style name="AppTheme">
    <item name="android:windowActionBarOverlay">true</item>
</style>
```

Additional Features
-----------
* `setDraggerView(View draggerView, int height)`    : Restric the drag area of the sliding panel to a specific view and height. Otherwise, the whole panel will be slidable and it will intercept all clicks.
* `setDraggerHeight(int height)`                    : Change the dragger height
* `setShadowDrawable(Drawable drawable)`            : Change the dragger shadow. No shadow displayed by default.
* `setShadowDrawable(Drawable drawable, int height)`: Change the dragger shadow with a custom height. No shadow displayed by default.
* `setContentHeight(int contentHeight)`             : Change the panel's content height to a custom one
* `setMaxContentHeightRatio(float ratio)`           : When using a custom height you can define the max height ratio of the view. 2f/3 by default, use 1.0f for full screen height.
* `collapse()` or `expand()`                        : Expand/Collapse the sliding panel
* `setInterceptingPanelEvents(boolean interceptTap)`: When using a custom dragger view, it might be usefull to make some views in the dragger clickable.

Requirements
-----------
* Install the support-v4-r13.jar to your own artifact repo as it is not available in the Maven repo's.
* Tested on Android 2.2+

Licence
-----------
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this work except in compliance with the License.
You may obtain a copy of the License in the LICENSE file, or at:

  [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Credits
-----------
* All credits goes to the [Umano Team](http://umanoapp.com): for the idea and the basic, but awesome implementation of this layout. You can see the original here: [AndroidSlidingUpPanel](https://github.com/umano/AndroidSlidingUpPanel/)
* This code is heavily based on the opened-sourced [SlidingPaneLayout](http://developer.android.com/reference/android/support/v4/widget/SlidingPaneLayout.html) component from the r13 of the Android Support Library. Thanks Android team!

=======

[![Build Status](https://secure.travis-ci.org/josomers/AndroidSlidingUpPanel.png)](https://travis-ci.org/josomers/AndroidSlidingUpPanel)


CHANGELOG.md
Change Log
==========

Version 1.1.1 *(2013-10-24)*
----------------------------

* Back to android-support-v4-r13.

Version 1.1 *(2013-10-24)*
----------------------------

 * Added Maven support with a sample and library module which makes it easy to test
 * Support for IntelliJ (Android Studio) & Eclipse
 * Refactored from `com.sothree.slidinguppaneldemo.SlidingUpPanelLayout` to a more generic name: `com.slidinguppanel.SlidingUpPanelLayout`. Inspired on the view naming conventions of `viewpagerindicator` for example.
 * Added support for a SlidingUpPanelLayout with a custom content height: `setContentHeight(int contentHeight)`
 * Added support for a SlidingUpPanellayout with a max content height, with a full height ratio (1.0f), or just a part of the screen: `setMaxContentHeightRatio(float ratio)`. Default set to a ratio of 2f/3.
 * Renamed `setPanelHeight(int height)` to `setDraggerHeight(int height)` as we have a panel with a dragger view
 * Renamed `setDragView(View dragView)` to `setDraggerView(View draggerView, int height)` where height is the height of the dragger view.
 * Simplified `collapse()` & `expand()` methods
 * Removed the SimplePanelSlideListener as it is not usefull
 * Set the DEFAULT_DRAGGER_HEIGHT to 40 (smaller than 68 as it was to high)
 * Renewed the drawScrim drawing as it was not always correct after making the view content dynamic, working with the setAlpha method on a Paint now which is much easier than swapping bits and bites around. DEFAULT_FADE_COLOR set to Color.BLACK (transparency not needed anymore)
* Enable/Disable tapping on the dragger layout (which makes it able to set buttons in it) (`setInterceptingPanelEvents(boolean interceptTap)`)
 * Added Robolectric and some very basic tests
 * Updated README.md with some new screenshots
 * Added travis build config

Version 1.0 *(2013-05-31)*
----------------------------

* Initial release.

slidinguppanel-sample/src/test/java/com/slidinguppanel/SampleTest.java
package com.slidinguppanel;

import org.junit.Test;

import static org.junit.Assert.assertNull;

public class SampleTest {

    @Test
    public void testSample() {
        assertNull(null); // not a real test, just added a default testcase.
    }

}


slidinguppanel/src/test/java/com/slidinguppanel/SlidingUpPanelLayoutTest.java
package com.slidinguppanel;

import android.content.Context;
import android.view.View;

import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

/**
 * @author : Jo Somers
 * Date: 22/07/13
 * Time: 16:48
 */

@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
public class SlidingUpPanelLayoutTest {

    private Context mContext;

    @Before
    public void setUp() {
        mContext = Robolectric.getShadowApplication().getApplicationContext();
    }

    @Test
    public void testInitialConfiguration() {
        SlidingUpPanelLayout slidingUpPanelLayout = new SlidingUpPanelLayout(mContext);
        assertNotNull(slidingUpPanelLayout);
        assertTrue(slidingUpPanelLayout.isSlideable());
        assertTrue(slidingUpPanelLayout.getVisibility() == View.VISIBLE);
        assertTrue(slidingUpPanelLayout.getMaxContentHeightRatio() == 2 / 3f);
        assertTrue(slidingUpPanelLayout.getContentHeight() == Integer.MAX_VALUE);
        assertTrue(slidingUpPanelLayout.isInterceptingPanelEvents());
        assertTrue(slidingUpPanelLayout.getDraggerHeight() == 40);
        assertTrue(slidingUpPanelLayout.getTopDistance(0) == 0);
        assertFalse(slidingUpPanelLayout.isActionUpEnabled());
    }

    @Test
    public void testCustomDraggerConfiguration() {
        SlidingUpPanelLayout slidingUpPanelLayout = new SlidingUpPanelLayout(mContext);
        slidingUpPanelLayout.setDraggerView(new View(mContext), 300);
        assertEquals(slidingUpPanelLayout.getDraggerHeight(), 300);

        slidingUpPanelLayout.setActionUpEnabled(true);
        assertTrue(slidingUpPanelLayout.isActionUpEnabled());
    }

    /**
     * Currently it is not possible to pass a test for expanding the Sliding Up Panel.
     * The method smoothSlideViewTo is not calling the methods (onViewPositionChanged) needed in the DragHelperCallback
     * setting the mSlideOffset to the right value.
     */
    @Ignore
    @Test
    public void testExpand() {
        // TODO
    }

    /**
     * Currently it is not possible to pass a test for collapsing the Sliding Up Panel.
     * The method smoothSlideViewTo is not calling the methods (onViewPositionChanged) needed in the DragHelperCallback
     * setting the mSlideOffset to the right value.
     */
    @Ignore
    @Test
    public void testCollapse() {
        //TODO
    }

}


slidinguppanel-sample/src/test/java/com/slidinguppanel/demo/test/SampleTest.java
package com.slidinguppanel.demo.test;

import com.slidinguppanel.demo.activities.MainActivity;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertNotNull;

/**
 * User: josomers
 * Date: 11/07/13
 * Time: 12:26
 */

@RunWith(RobolectricTestRunner.class)
public class SampleTest {

    @Test
    public void testSample() {
        MainActivity mainActivity = Robolectric.buildActivity(MainActivity.class).create().get();
        assertNotNull(mainActivity); // not a real test, just added a default testcase.
    }

}


slidinguppanel-sample/src/main/java/com/slidinguppanel/demo/activities/MainActivity.java
package com.slidinguppanel.demo.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.slidinguppanel.demo.R;

/**
 * User: josomers
 * Date: 18/07/13
 * Time: 14:21
 */
public class MainActivity extends Activity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button customButton = (Button) findViewById(R.id.custom_buttons);
        customButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), CustomPanelButtonsActivity.class));
            }
        });

        Button customTextView = (Button) findViewById(R.id.custom_textviews);
        customTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), CustomPanelTextViewsActivity.class));
            }
        });

        Button fullscreen = (Button) findViewById(R.id.fullscreen);
        fullscreen.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), FullscreenActivity.class));
            }
        });
    }

}


slidinguppanel-sample/src/main/java/com/slidinguppanel/demo/activities/CustomPanelButtonsActivity.java
package com.slidinguppanel.demo.activities;

import android.annotation.TargetApi;
import android.app.Activity;
import android.os.Build;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;

import com.slidinguppanel.SlidingUpPanelLayout;
import com.slidinguppanel.demo.R;

public class CustomPanelButtonsActivity extends Activity {

    private SlidingUpPanelLayout mSlidingUpPanelLayout;
    private final float mDefaultRatio = 2f / 3; // two-thirds of the screen height - ratio.

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_custom_panel_buttons);

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.HONEYCOMB) {
            getActionBar().setHomeButtonEnabled(true);
            getActionBar().setDisplayHomeAsUpEnabled(true);
        }

        final LinearLayout panel = (LinearLayout) findViewById(R.id.panel);

        mSlidingUpPanelLayout = (SlidingUpPanelLayout) findViewById(R.id.slidinguppanel);
        mSlidingUpPanelLayout.setInterceptingPanelEvents(true);
        mSlidingUpPanelLayout.setDraggerView(panel, getResources().getDimensionPixelSize(R.dimen.panel_height_custom));
        mSlidingUpPanelLayout.setShadowDrawable(getResources().getDrawable(R.drawable.above_shadow));
        mSlidingUpPanelLayout.setMaxContentHeightRatio(mDefaultRatio);

        final Button btnPanel1 = (Button) findViewById(R.id.expand);
        btnPanel1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSlidingUpPanelLayout.expand();
            }
        });

        final Button btnPanel2 = (Button) findViewById(R.id.collapse);
        btnPanel2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSlidingUpPanelLayout.collapse();
            }
        });

        final Button button1 = (Button) findViewById(R.id.button1);
        button1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSlidingUpPanelLayout.setMaxContentHeightRatio(1.0f);
                mSlidingUpPanelLayout.expand();
            }
        });

        final Button button2 = (Button) findViewById(R.id.button2);
        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                buttonPressed(200);
            }
        });

        final Button button3 = (Button) findViewById(R.id.button3);
        button3.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                buttonPressed(700);
            }
        });

        final Button button4 = (Button) findViewById(R.id.button4);
        button4.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSlidingUpPanelLayout.setMaxContentHeightRatio(mDefaultRatio);
                mSlidingUpPanelLayout.expand();
            }
        });

        final Button button5 = (Button) findViewById(R.id.button5);
        button5.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSlidingUpPanelLayout.resetContentHeight();
                mSlidingUpPanelLayout.expand();
            }
        });
    }

    private void buttonPressed(int height) {
        mSlidingUpPanelLayout.setContentHeight(height);
        mSlidingUpPanelLayout.expand();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }

        return super.onOptionsItemSelected(item);
    }
}

slidinguppanel-sample/src/main/java/com/slidinguppanel/demo/activities/CustomPanelTextViewsActivity.java
package com.slidinguppanel.demo.activities;

import android.annotation.TargetApi;
import android.app.Activity;
import android.os.Build;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.slidinguppanel.SlidingUpPanelLayout;
import com.slidinguppanel.demo.R;

public class CustomPanelTextViewsActivity extends Activity {

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_custom_panel_textviews);

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.HONEYCOMB) {
            getActionBar().setHomeButtonEnabled(true);
            getActionBar().setDisplayHomeAsUpEnabled(true);
        }

        final View panel = findViewById(R.id.panel);

        final SlidingUpPanelLayout slidingUpPanelLayout = (SlidingUpPanelLayout) findViewById(R.id.slidinguppanel);
        slidingUpPanelLayout.setInterceptingPanelEvents(true);
        slidingUpPanelLayout.setDraggerView(panel, getResources().getDimensionPixelSize(R.dimen.panel_height_custom));
        slidingUpPanelLayout.setShadowDrawable(getResources().getDrawable(R.drawable.above_shadow), 5);
        slidingUpPanelLayout.setMaxContentHeightRatio(1.0f);

        final TextView btnPanel1 = (TextView) findViewById(R.id.expand);
        btnPanel1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                slidingUpPanelLayout.expand();
            }
        });

        final TextView btnPanel2 = (TextView) findViewById(R.id.collapse);
        btnPanel2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                slidingUpPanelLayout.collapse();
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }

        return super.onOptionsItemSelected(item);
    }
}

slidinguppanel-sample/src/main/java/com/slidinguppanel/demo/activities/FullscreenActivity.java
package com.slidinguppanel.demo.activities;

import android.annotation.TargetApi;
import android.app.Activity;
import android.os.Build;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;

import com.slidinguppanel.SlidingUpPanelLayout;
import com.slidinguppanel.demo.R;

/**
 * User: josomers
 * Date: 18/07/13
 * Time: 14:10
 */
public class FullscreenActivity extends Activity {

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().requestFeature(Window.FEATURE_ACTION_BAR_OVERLAY);
        setContentView(R.layout.activity_fullscreen);

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.HONEYCOMB) {
            getActionBar().setHomeButtonEnabled(true);
            getActionBar().setDisplayHomeAsUpEnabled(true);
        }

        final SlidingUpPanelLayout slidingUpPanelLayout = (SlidingUpPanelLayout) findViewById(R.id.slidinguppanel);
        slidingUpPanelLayout.setInterceptingPanelEvents(true);
        slidingUpPanelLayout.setDraggerHeight(getResources().getDimensionPixelSize(R.dimen.panel_height_full));
        slidingUpPanelLayout.setShadowDrawable(getResources().getDrawable(R.drawable.above_shadow));
        slidingUpPanelLayout.setMaxContentHeightRatio(1.0f);

        slidingUpPanelLayout.setPanelSlideListener(new SlidingUpPanelLayout.PanelSlideListener() {
            @Override
            public void onPanelSlide(View view, float slideOffset) {
                if (Build.VERSION.SDK_INT > Build.VERSION_CODES.HONEYCOMB) {
                    if (slideOffset < 0.2) {
                        if (getActionBar().isShowing()) {
                            getActionBar().hide();
                        }
                    } else {
                        if (!getActionBar().isShowing()) {
                            getActionBar().show();
                        }
                    }
                }
            }

            @Override
            public void onPanelCollapsed(View view) {
            }

            @Override
            public void onPanelExpanded(View view) {
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }

        return super.onOptionsItemSelected(item);
    }

}
